### ISSUE TITLE
Discovery document does generate wrong paths because BaseUrlMiddleware does not respect PublicOrigin correctly

### ISSUE THREAD

**Issue by gingters** [IdentityServer/IdentityServer4#3009]
When deploying IdentityServer in ServiceFabric, ServiceFabric acts as a reverse proxy and assigns random, unique Urls to internal services.

The public Url of our service is `https://publichost:publicport/identity/`, however service fabric internally accesses it as `https://internalhost:internalport/{guid}/{guid}/`, with both guids being dynamically assigned at service startup time.

We have set the public url of the fabric host to both `IssuerUrl` as well as `PublicOrigin`. However, in the discovery document the urls to the endpoints are generated as `https://publichost:publicport/identity/{guid}/{guid}/.well-known/...`, exposing the pure internal url's to the outside. The endpoints are therefore not accessible to the outside.

The `BaseUrlMiddleware` in IdentityServer does not provide a way to override the path and change the way the urls are generated except for the host part.
There needs to be a way to prevent IdentityServer to add these two guids from the current (internal) request path being added to the generated urls in the discovery document i.e. setting a public base path in addition to public origin, which would override the `request.PathBase` in the `BaseUrlMiddleware`, just like the public origin setting does with the external host.

**Comment by leastprivilege** [comment]
We'll have a look. Thanks for reporting.

**Comment by brockallen** [comment]
I had a look and I think your best bet is to write your own middleware in front to fix the path to be what you need/want. Our public origin is just to set the orgin (hence the name) not the path (as the origin is different than the URL and path).

**Comment by gingters** [comment]
Hrm. I tried that, but that causes other issues. The `BaseUrlMiddleware` uses the `request.Context.PathBase` to set the `IdentityServerBasePath` variable on the context.

When I put a Middleware in front of IdentityServer to change the `request.Context.PathBase` to not include the purely internal path, then the UI Controllers (based and adopted from the Quickstart UI) aren't capable of building the correct redirect url's to the authorize endpoint after login.
And the relative links to CSS and JS aren't created correctly, too.

Writing that.. maybe if I keep the old PathBase somewhere and restore that in another Middleware right after IdentityServer…

I will investigate further and keep you informed with what I figured out.
